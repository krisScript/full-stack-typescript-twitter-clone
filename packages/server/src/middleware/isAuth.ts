import { verify } from 'jsonwebtoken';
import { Request, Response, NextFunction } from 'express';
import { CustomError, errors } from '@utilities/CustomError';

const isAuth = (req: Request, _: Response, next: NextFunction): void => {
  const { SECRET } = process.env;
  const authHeader = req.get('Authorization');
  const { status, message } = errors.Unauthorized;
  const error = new CustomError(status, message);
  if (!authHeader) {
    throw error;
  }
  const token = authHeader.split(' ')[1];
  const decodedToken = verify(token, SECRET);
  // @ts-ignore
  const { userId } = decodedToken;
  if (!decodedToken || !userId) {
    throw error;
  }
  req.userId = userId;
  next();
};
export default isAuth;
