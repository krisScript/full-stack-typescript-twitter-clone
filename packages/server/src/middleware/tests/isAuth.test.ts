import mongoose from 'mongoose';
import httpMocks from 'node-mocks-http';
import jwt from 'jsonwebtoken';
import isAuth from '@customMiddleware/isAuth';

describe('isAuth', (): void => {
  it(`should add userId to req when the authorization header is valid`, async (): Promise<
    void
  > => {
    expect.assertions(2);
    const nextMock = jest.fn();
    const userId = mongoose.Types.ObjectId();
    const secret = process.env.SECRET;
    const token = jwt.sign(
      {
        userId,
      },
      secret,
      { expiresIn: '1h' },
    );
    const reqMock = httpMocks.createRequest({
      method: 'POST',
      url: '/',
      headers: {
        Authorization: `Bearer ${token}`,
      },
    });
    const resMock = httpMocks.createResponse();
    isAuth(reqMock, resMock, nextMock);
    expect(nextMock).toBeCalledTimes(1);
    expect(reqMock.userId).toMatch(userId.toString());
  });
  it('should throw an error with a status of 401: Unauthorized when there is no authorization header or its contents are invalid', async (): Promise<
    void
  > => {
    expect.assertions(1);
    const nextMock = jest.fn();
    const reqMock = httpMocks.createRequest({
      method: 'POST',
      url: '/',
    });
    const resMock = httpMocks.createResponse();
    expect((): void =>
      isAuth(reqMock, resMock, nextMock),
    ).toThrowErrorMatchingSnapshot();
  });
});
