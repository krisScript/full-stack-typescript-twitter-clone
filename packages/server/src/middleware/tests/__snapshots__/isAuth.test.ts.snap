// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`isAuth should throw an error with a status of 401: Unauthorized when there is no authorization header or its contents are invalid 1`] = `"Authentication credentials not valid or missing"`;
